client/session.go

```diff
@@ -781,24 +837,68 @@ func (s *session) writeAttempt(
 		return errSessionStateNotOpen
 	}
 
+	var (
+		enqueued int32
+		majority = atomic.LoadInt32(&s.majority)
+	)
+
+	// NB(prateek): We retain an individual copy of the namespace, ID per
+	// writeState, as each writeState tracks the lifecycle of it's resources in
+	// use in the various queues. Tracking per writeAttempt isn't sufficient as
+	// we may enqueue multiple writeStates concurrently depending on retries
+	// and consistency level checks.
+	nsID := s.idPool.Clone(namespace)
+	tsID := s.idPool.Clone(id)
+	var tagEncoder serialize.Encoder
+	if wType == taggedWriteAttemptType {
+		tagEncoder = s.tagEncoderPool.Get()
+		if err := tagEncoder.Encode(inputTags); err != nil {
+			tagEncoder.Finalize()
+			s.RUnlock()
+			return err
+		}
+	}
+
+	var op writeOp
+	switch wType {
+	case untaggedWriteAttemptType:
+		wop := s.writeOperationPool.Get()
+		wop.namespace = nsID
+		wop.shardID = s.topoMap.ShardSet().Lookup(tsID)
+		wop.request.ID = tsID.Data().Get()
+		wop.request.Datapoint.Value = value
+		wop.request.Datapoint.Timestamp = timestamp
+		wop.request.Datapoint.TimestampTimeType = timeType
+		wop.request.Datapoint.Annotation = annotation
+		op = wop
+	case taggedWriteAttemptType:
+		wop := s.writeTaggedOperationPool.Get()
+		wop.namespace = nsID
+		wop.shardID = s.topoMap.ShardSet().Lookup(tsID)
+		wop.request.ID = tsID.Data().Get()
+		encodedTagBytes, ok := tagEncoder.Data()
+		if !ok {
+			return errUnableToEncodeTags
+		}
+		wop.request.EncodedTags = encodedTagBytes.Get()
+		wop.request.Datapoint.Value = value
+		wop.request.Datapoint.Timestamp = timestamp
+		wop.request.Datapoint.TimestampTimeType = timeType
+		wop.request.Datapoint.Annotation = annotation
+		op = wop
+	default:
+		// should never happen
+		return errUnknownWriteAttemptType
```

Comment:
Also missing `RUnlock()` here too.

Reply:
addressed by the `...WithRLock` refactor